src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts

```diff
@@ -86,7 +91,217 @@ class FormatOnSaveParticipant implements IStoredFileWorkingCopySaveParticipant {
 				return [];
 			}));
 
-			await this.bulkEditService.apply(/* edit */allCellEdits.flat(), { label: localize('label', "Format Notebook"), code: 'undoredo.formatNotebook', });
+			await this.bulkEditService.apply(/* edit */allCellEdits.flat(), { label: localize('formatNotebook', "Format Notebook"), code: 'undoredo.formatNotebook', });
+
+		} finally {
+			progress.report({ increment: 100 });
+			disposable.dispose();
+		}
+	}
+}
+
+class TrimWhitespaceParticipant implements IStoredFileWorkingCopySaveParticipant {
+
+	constructor(
+		@IConfigurationService private readonly configurationService: IConfigurationService,
+		@IEditorService private readonly editorService: IEditorService,
+		@ITextModelService private readonly textModelService: ITextModelService,
+		@IBulkEditService private readonly bulkEditService: IBulkEditService,
+	) { }
+
+	async participate(workingCopy: IStoredFileWorkingCopy<IStoredFileWorkingCopyModel>, context: { reason: SaveReason }, progress: IProgress<IProgressStep>, _token: CancellationToken): Promise<void> {
+		if (this.configurationService.getValue<boolean>('files.trimTrailingWhitespace')) {
+			await this.doTrimTrailingWhitespace(workingCopy, context, progress);
+		}
+	}
+
+	private async doTrimTrailingWhitespace(workingCopy: IStoredFileWorkingCopy<IStoredFileWorkingCopyModel>, context: { reason: SaveReason }, progress: IProgress<IProgressStep>) {
+		if (!workingCopy.model || !(workingCopy.model instanceof NotebookFileWorkingCopyModel)) {
+			return;
+		}
+
+		const disposable = new DisposableStore();
+		const notebook = workingCopy.model.notebookModel;
+
+		const cursors: Position[] = [];
+		let viewCell: ICellViewModel | undefined = undefined;
+
+		// autosave -- don't trim entire line, only up to cursor, so need to track position of cursor(s)
+		if (context.reason === SaveReason.AUTO) {
+			viewCell = getNotebookViewCell(this.editorService);
+			if (!viewCell) {
+				return;
+			}
+			const selections = viewCell.getSelections();
+			for (const sel of selections) {
+				if (viewCell.model.textModel) {
+					cursors.push(new Position(sel.selectionStartLineNumber, sel.startColumn));
```

Comment:
If we are tracking the cursor it's `positionLineNumber` and `positionColumn`, not start*

Reply:
We may want to follow https://github.com/microsoft/vscode/blob/7151639826490c7af63befbee10a0c229c47b020/src/vs/workbench/contrib/codeEditor/browser/saveParticipants.ts#L63-L71